[supply] add new SUPPLY_UPLOAD_MAX_RETRIES env var to attempt to solve failed Google API calls

joshdholtz opened the pull request:
### Motivation and Context

Somewhat fixes #21507

Applying the patch made by @AliSoftware mentioned https://github.com/fastlane/fastlane/issues/21507#issuecomment-1721417442

### Description

Added new `SUPPLY_UPLOAD_MAX_RETRIES` env variable to retry failed Google API requests to attempt to "fix" uploads from erroring

```sh
SUPPLY_UPLOAD_MAX_RETRIES=5 fastlane your_lane
```

joshdholtz committed:
[supply] add new SUPPLY_UPLOAD_MAX_RETRIES env var to attempt to solve failed Google API calls

AliSoftware commented on supply/spec/client_spec.rb:
```diff
@@ -8,23 +8,47 @@
+          to_return(status: 403, body: '{"error":{"message":"Ensure project settings are enabled."}}', headers: { 'Content-Type' => 'application/json' })
+
+        expect(UI).to receive(:error).with("Google Api Error: Invalid request - Ensure project settings are enabled. - Retrying...").exactly(5).times
+        expect(UI).to receive(:user_error!).with("Google Api Error: Invalid request - Ensure project settings are enabled.").once
```
One thing that I got bitten by in the past is that when you stub `UI.user_error!` with `expect` or `allow`, that means that the original method won't be called… which means that it won't `raise`.

All that means that the caller of `call_google_api`, in your case `client.upload_image`, won't be interrupted by the `Fastlane::Error` raised by `user_error!` in the context of running tests, while it will in the real world. Depending on the implementation of `client.upload_image(…)`, that could have side effects if that implementation counts on the fact that `call_google_api` can raise to stop early and not run the rest of the code of that method on error, as that would mean that the fact that we stubbed `user_error!` would make the method under test behave differently in test environment vs in reality.

I think the easiest way to solve this would be to replace your `expect(UI).to receive(:user_error!)` with wrapping your call to `client.upload_image` within `expect { … }.to raise_error(Fastlane::Error, "Google Api Error: Invalid request – Ensure project settings are enabled."`. That way you'd let `user_error!` raise and interrupt the code of `upload_image` mid-air even in the context of testing.

joshdholtz replied:
@AliSoftware That expect on `UI.user_error!` was already there 🤷 But... would be adding `.and_call_original` work? 🤔

AliSoftware replied:
Well that would, except that I _think_ `rubocop` might suggest you to only call `.and_call_original` on `allow(…)`, not on `expect(…)`

So you could `allow(UI).to receive(:user_error!).with(…).and_call_original`… but if you do that and even if you wanted to still keep the `expect(UI).to receive(:user_error!).once` to assert on it only being called once and not more, you'll still need to wrap your `client.upload_image` within `expect { … }.to raise_error(…)` to avoid the test from crashing on the raised error and instead catch (and assert) it.

So at this point since you'll still need the `expect { … } .to raise_error`, I'm not sure that `and_call_original` adds much value, instead of not stubbing `UI.user_error!` at all?

AliSoftware committed:
Rewrite expectation to ensure SUT is interrupted by raised error

Otherwise if we just `expect(UI).to receive(:user_error!).with(…)` that will mock the `UI.user_error` call in the method being tested... and thus not make it `raise` anymore, which the rest of the implementation of that method will continue executing when run in the context of mocked tests while it would not in the real world, which means we would be testing a different behavior of the method

AliSoftware reviewed (approved):
@joshdholtz I hope you don't mind, I've pushed a commit to apply the suggested change I was talking about for the spec 🙂

joshdholtz commented:
@AliSoftware Thank you! I was at the lake yesterday and didn't have my computer on me. Just getting up with the family now so will get this merged and a new release made 💪
